great_expectations/expectations/metrics/util.py

```diff
@@ -896,3 +898,53 @@ def is_valid_continuous_partition_object(partition_object):
         and np.all(np.diff(partition_object["bins"]) > 0)
         and np.allclose(np.sum(comb_weights), 1.0)
     )
+
+
+def sql_statement_with_post_compile_to_string(
+    engine: "sqlalchemy.engine.Engine", select_statement: "sqlalchemy.sql.Select"
+) -> str:
+    """
+    Util method to compile SQL select statement with post-compile parameters into a string. Logic lifted directly
+    from sqlalchemy documentation.
+
+    https://docs.sqlalchemy.org/en/14/faq/sqlexpressions.html#rendering-postcompile-parameters-as-bound-parameters
+
+    Used by _sqlalchemy_map_condition_index() in map_metric_provider to build query that will allow you to
+    return unexpected_index_values.
+
+    Args:
+        engine (sqlalchemy.engine.Engine): Sqlalchemy engine used to do the compilation.
+        select_statement (sqlalchemy.sql.Select): Select statement to compile into string.
+    Returns:
+        String representation of select_statement
+    """
+    compiled = select_statement.compile(
+        engine, compile_kwargs={"render_postcompile": True}
+    )
+    params = (repr(compiled.params[name]) for name in compiled.positiontup)
+    query_as_string = re.sub(r"\?", lambda m: next(params), str(compiled))
```

Comment:
The test for this could just be for this method to make sure we get the expected query, no need to set up a data connector, validator etc. There are some examples in the splitting / sampling tests, e.g. in test_sqlalchemy_execution_engine_sampling.py this fixture: dialect_name_to_sql_statement and this test test_sample_using_limit_builds_correct_query_where_clause_none